File: test/더맵게.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
//효율성은 조만간...

// K <= 섞 = 가장 안매운거 + (두 번째로 안매운거 * 2)
// mix를 vector 뒤에 넣고 정렬함.
int solution(vector<int> scoville, int K) {
    int answer = 0;
    int mix = 0;

    /*if(scoville.size() == 1){
        mix = scoville[0];
    }*/
    
    while (1){
        
        //배열에 원소 하나만 남았을때
        if(scoville.size() == 1){
            if(scoville[0] < K) //스코빌 지수가 K보다 작으면
                return -1;
            else 
                return answer;
        } 
        else{
            sort(scoville.begin(), scoville.end());
            if(scoville[0] >= K)    //정렬하고 비교!!
                return answer;
            mix = scoville[0] + (scoville[1] * 2);
            scoville.push_back(mix);
            scoville.erase(scoville.begin());
            scoville.erase(scoville.begin());
            answer++;
        }
    }
    
    return answer;
}

File: python_algorithm/힙/더맵게.py
# 스코빌 지수를 K 이상으로 만들기 위해서 스코빌 지수가 가장 낮은 두 개의 음식을 
# 섞은 = 가장 맵지 않은 + (두 번째로 맵지 않은 음식의 스코빌 지수 * 2)
# 모든 스코빌 지수가 K 이상이 될 때까지 섞음
# 최소 횟수를 return 하도록 
# 힙 정렬.... 최소 트리로 하면...되지 않을까? 
import heapq    #heapq 라는것이 존재한다.... 참나...
scoville = [1, 2, 3, 9, 10, 12]
K = int(input())

def solution(scoville, K):
    mix = 0
    answer = 0
    heapq.heapify(scoville) #scoville 배열을 heap으로 바꾼다. 
    while scoville != []:
        if len(scoville) == 1:
            if(scoville[0] < K):
                return -1
            else:
                return answer
        else:
            h1 = heapq.heappop(scoville)    #heappop후 자동으로 힙정렬을 해준다.
            if(h1 >= K):
                return answer
            h2 = heapq.heappop(scoville)
            heapq.heappush(scoville, h1 + h2*2)
            answer += 1

print(solution(scoville, K))




